lightgbm/src/main/scala/com/microsoft/ml/spark/lightgbm/dataset/DatasetAggregator.scala

```diff
@@ -0,0 +1,564 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.ml.spark.lightgbm.dataset
+
+import java.util.concurrent.atomic.AtomicLong
+
+import com.microsoft.ml.spark.io.http.SharedSingleton
+import com.microsoft.ml.spark.lightgbm.{ColumnParams, SharedState}
+import com.microsoft.ml.spark.lightgbm.swig.{BaseSwigArray, ChunkedArray, DoubleChunkedArray, DoubleSwigArray,
+  FloatChunkedArray, FloatSwigArray, IntChunkedArray, IntSwigArray}
+import com.microsoft.ml.spark.lightgbm.dataset.DatasetUtils.{addFeaturesToChunkedArray, addGroupColumnRow,
+  addInitScoreColumnRow, getArrayType, getRowAsDoubleArray}
+import com.microsoft.ml.spark.stages.LocalAggregator
+import org.apache.spark.ml.linalg.{DenseVector, SparseVector}
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.types.StructType
+
+import scala.collection.mutable.ListBuffer
+
+private[lightgbm] object ChunkedArrayUtils {
+  def copyChunkedArray[T: Numeric](chunkedArray: ChunkedArray[T],
+                                   mainArray: BaseSwigArray[T],
+                                   threadRowStartIndex: Long,
+                                   chunkSize: Long): Unit = {
+    val num = implicitly[Numeric[T]]
+    val defaultVal = num.fromInt(-1)
+    // Copy in parallel on each thread
+    // First copy full chunks
+    val chunkCount = chunkedArray.getChunksCount() - 1
+    for (chunk <- 0L until chunkCount) {
+      for (inChunkIdx <- 0L until chunkSize) {
+        mainArray.setItem(threadRowStartIndex + chunk * chunkSize + inChunkIdx,
```

Comment:
it seems that this is not a bottleneck based on benchmarking, but it can definitely be improved and it will be when this will be moved inside the lightgbm native code, but not in this PR